lib/serde-fallback.js

```diff
@@ -2,691 +2,743 @@
-    undefined,
-    null: null,
-    true: true,
-    false: false,
-  };
+      if (!encountered.dot && encountered.exponent) {
+        isFloatExponent = true;
+      } else if (encountered.dot && isFloatExponent) {
+        this.throwError('Invalid format');
+      }
+      number += this.advance();
+    }
 
-  if (matching.hasOwnProperty(identifier)) {
-    return matching[identifier];
-  } else {
-    return this.throwUnexpected();
+    return parseFloat(number);
+  }
+
+  // Check whether a part of a decimal number that can be used only once
+  // has been encountered before and throw SyntaxError if it indeed has.
+  //   encounterContext - an object with boolean flags
+  //   name - name of the key of the object which value represents the
+  //     entity being checked
+  //   character - a character to match against the lookahead character
+  //   altCharacter - optional alternative character related to the
+  //     same enitity
+  //
+  checkNumberPartEncountered(encounterContext, name, character, altCharacter) {
+    const look = this.lookahead();
+    if (look === character || (altCharacter && look === altCharacter)) {
+      if (encounterContext[name]) {
+        this.throwUnexpected();
+      }
+      encounterContext[name] = true;
+    }
   }
-};
 
-// Parse a single-quoted or double-quoted string
-//
-Parser.prototype.parseString = function() {
-  this.skipClutter();
+  // Parse an integer number in binary, octal or hexadecimal representation.
+  // The prefix is not included and must be parsed before invocation.
+  // The function may return NaN when the number is incorrect.
+  //   base - 2, 8 or 16
+  //
+  parseMachineInteger(base) {
+    const self = this;
+    const checkers = {
+      2: self.isBinaryDigit,
+      8: self.isOctalDigit,
+      16: self.isHexadecimalDigit,
+    };
+    const checkDigit = checkers[base];
+
+    let number = '';
+    while (checkDigit(this.lookahead())) {
+      number += this.advance();
+    }
 
-  const quoteStyle = this.lookahead();
-  if (!this.isQuoteCharacter(quoteStyle)) {
-    this.throwExpected('String');
+    return parseInt(number, base);
   }
 
-  let string = '';
-  let escapeMode = false;
+  // Parse null, undefined, true or false
+  //
+  parseIdentifier() {
+    this.skipClutter();
 
-  this.advance();
+    let identifier = '';
+    while (this.isLetter(this.lookahead())) {
+      identifier += this.advance();
+    }
 
-  while (escapeMode || this.lookahead() !== quoteStyle) {
-    const look = this.advance();
+    if (identifier === 'NaN') {
+      return NaN;
+    }
 
-    if (escapeMode) {
-      const controlCharacters = {
-        b: '\b',
-        f: '\f',
-        n: '\n',
-        r: '\r',
-        t: '\t',
-        v: '\v',
-      };
+    if (identifier === 'Infinity') {
+      return Infinity;
+    }
 
-      const controlCharacter = controlCharacters[look];
+    const matching = {
+      undefined,
+      null: null,
+      true: true,
+      false: false,
+    };
```

Comment:
There is no need to check for `NaN` and `Infinity` above specifically, just add two lines to this object:
```suggestion
      NaN,
      Infinity,
    };
```